introduction.Rmd

```diff
@@ -9,27 +9,185 @@ library("dplyr")
+
+Changing the context of evaluation is useful for three main purposes:
+
+* To promote data frames to **full blown scopes**, where columns are exposed as named objects.
+
+* To execute your R code in a **foreign environment**. For instance, dbplyr translates ordinary dplyr pipelines to SQL queries.
+
+* To implement **special rules** for ordinary R operators. For instance, selection functions such as `dplyr::select()` or `tidyr::gather()` implement specific behaviours for `c()`, `:` and `-`.
+
+
+## Data masking
+
+Of these goals, the promotion of data frames is the most important because data is often the most relevant context for data analysts. We believe that R and the tidyverse are [human-centered](https://principles.tidyverse.org/unifying-principles.html#human-centered) in big part because the data frame is available for direct use in computations, without syntax and boilerplate getting in the way. Formulas for statistical models are a prime example of human-centered syntax in R. Data masking and special operator rules make formulas an intuitive interface for model specification.
+
+When the contents of the data frame are temporarily promoted as first class objects, we say the data **masks** the workspace:
 
 ```{r, eval = FALSE}
 library("dplyr")
 
-# Masking data
-starwars %>%
-  filter(
-    height < 200,
-    gender == "male"
-  )
+starwars %>% filter(
+  height < 200,
+  gender == "male"
+)
+```
+
+Compare to the equivalent subsetting code where it is necessary to be explicit about where the columns come from:
 
-# Qualified data
-starwars[starwars$height < 200 & starwars$gender == "male", ]
+```{r, eval = FALSE}
+index <-
+  !is.na(starwars$height) & !is.na(starwars$gender) &
+  starwars$height < 200 & starwars$gender == "male"
+
+starwars[index, ]
 ```
 
-Data masking makes it easy and natural to read and write data manipulation code, but it has a flip side. It is easier to refer to objects in the mask when you know their names in advance, but it is harder when the names are unknown at the time of writing. In particular, it is harder to refer to columns whose names are stored in variables or passed as function arguments.
+Data masking is natural for many users because it reduces boilerplate and results in code that maps more directly to how they think about data manipulation problems. This is why many R beginners are tempted to use `attach()`. This function attaches a data frame to the search path in much the same way that `library()` attaches exported objects of a package to make it easier to use interactively. However `attach()` is not a recommended way of using a data mask for the same reason that global options are frown on — it makes your code less reproducible because of the permanent scope of these effects. On the other hand, data masking functions have only temporary effects, making them a good choice for both interactive and programmatic uses.
```

Comment:
~~many users~~

And I think mentioning `attach()` here is just a distraction. Assume people are basically on board with tidy eval when reading this intro. Maybe we should have a later chapter called "tidy eval for sceptics"?